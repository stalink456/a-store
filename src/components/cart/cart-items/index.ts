export { CartItems } from "./cart-items";
