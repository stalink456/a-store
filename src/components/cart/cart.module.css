@import "../../css/variables.css";
@import "../../css/animation.css";

.cart__wrapper {
  position: fixed;
  top: 100px;
  right: 20px;
  border-radius: 50px;
  background-color: var(--logo-color);
  composes: fade_scale from global;
  z-index: 1;
}

.cart__icon {
  color: var(--white-color);
  font-size: 54px;
  padding: 20px;
}

.cart__icon_text {
  position: absolute;
  width: 35px;
  text-align: center;
  color: var(--white-color);
  background-color: var(--black-color);
  border-radius: 25px;
  padding: 5px;
  top: 70%;
  right: 0;
}

.cart__icon_tooltip {
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
  top: 25%;
  right: 60%;
  width: 120px;
  height: 40px;
  background-color: var(--black-color);
  color: var(--white-color);
  transform: translate(-50%);
  border-radius: 10px;
  visibility: hidden;
  opacity: 0;
  transition: opacity 1s ease-in-out;
  composes: tooltip_visible from global;
}

.cart__icon_tooltip::before {
  content: "";
  position: absolute;
  left: 110%;
  top: 50%;
  transform: translate(-50%, -50%);
  border: 15px solid;
  border-color: var(--border-color) var(--border-color) var(--border-color)
    var(--black-color);
}

.cart__icon_tooltip__text {
  margin: 0;
}

.cart__wrapper:hover .cart__icon_tooltip {
  visibility: visible;
  opacity: 1;
}

.cart__wrapper:hover {
  transform: scale(1.1);
}

@media (max-width: 960px) {
  .cart__wrapper {
    top: 80vh;
  }
}

@media (max-width: 640px) {
  .cart__icon {
    font-size: 40px;
    padding: 10px;
  }

  .cart__icon_text {
    width: 30px;
    padding: 3px;
    left: 60%;
    top: 60%;
  }

  .cart__icon_tooltip {
    top: 20%;
    right: 40%;
  }
}
