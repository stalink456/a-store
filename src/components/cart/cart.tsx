import React from "react";
import { IconButton } from "@alfalab/core-components/icon-button";
import { AiOutlineShopping } from "react-icons/ai";
import { Typography } from "@alfalab/core-components/typography";
import { Space } from "@alfalab/core-components/space";
import { useSidePanel } from "hooks/use-side-panel";
import { CartPanel } from "./cart-panel/cart-panel";
import { useCart } from "hooks/use-cart";

import styles from "./cart.module.css";

export const Cart: React.FC = React.memo(() => {
  const { cartItemsLength, totalCount, cartTotalPrice, cartRef } = useCart();
  const { open, handleOpenMenu } = useSidePanel(cartRef);

  const renderCart = () => {
    return cartItemsLength ? (
      <div key={totalCount} className={styles.cart__wrapper}>
        <IconButton
          className={styles.cart__icon}
          icon={AiOutlineShopping}
          size="s"
          onClick={handleOpenMenu}
          data-testid="cart-icon"
        />
        <Typography.Text className={styles.cart__icon_text} tag="p">
          {totalCount}
        </Typography.Text>
        <Space className={styles.cart__icon_tooltip}>
          <Typography.Text
            className={styles.cart__icon_tooltip__text}
            tag="p"
            weight="bold"
            view="primary-large"
          >
            = {cartTotalPrice} â‚½
          </Typography.Text>
        </Space>
      </div>
    ) : null;
  };

  return (
    <div className={styles.cart}>
      {renderCart()}
      <CartPanel ref={cartRef} open={open} handleOpenMenu={handleOpenMenu} />
    </div>
  );
});
