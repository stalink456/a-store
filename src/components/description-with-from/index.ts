export { DescriptionWithForm } from './description-with-form';
