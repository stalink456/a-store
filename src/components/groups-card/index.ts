export { GroupsCard } from './groups-card';
