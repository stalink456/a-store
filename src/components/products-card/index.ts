export { ProductsCard } from "./products-card";
