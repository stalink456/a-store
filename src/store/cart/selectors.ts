import { ApplicationState } from '..';

export const cartSelector = (state: ApplicationState) => state.cart;

export const cartItemsSelector = (state: ApplicationState) =>
  cartSelector(state).items;
export const cartItemsLengthSelector = (state: ApplicationState) =>
  cartItemsSelector(state).length;
export const cartTotalPriceSelector = (state: ApplicationState) =>
  cartSelector(state).cartTotalPrice;
