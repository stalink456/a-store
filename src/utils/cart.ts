import { CartType } from "types";

export const calcTotalPrice = (items: CartType[]) => {
  return items.reduce((sum, obj) => {
    return obj.price * obj.totalCount + sum;
  }, 0);
};

export const getCartFromLS = () => {
  const data = localStorage.getItem("cart");
  const items: CartType[] = data ? JSON.parse(data) : [];
  const cartTotalPrice = calcTotalPrice(items);

  return {
    items: items,
    cartTotalPrice,
  };
};

export const calculateTotalCount = (items: CartType[]) => {
  return items.reduce((sum: number, { totalCount }) => sum + totalCount, 0);
};

export const generateID = (
  color: string | null,
  size: string | null,
  stickerName: string | null
) => {
  return [color, size, stickerName].filter((value) => value !== null).join("");
};
