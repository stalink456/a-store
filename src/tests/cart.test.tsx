import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Cart } from '../components/cart';
import * as items from './mocks/get-cart-items-200.json';
import * as cartLength from './mocks/get-cart-length-200.json';
import * as totalCount from './mocks/get-total-count-200.json';
import * as totalPrice from './mocks/get-total-price-200.json';

const mockCartItems = items['items'];
const mockedLength = cartLength['length'];
const mockedCount = totalCount['count'];
const mockedPrice = totalPrice['price'];

jest.mock('hooks/use-cart', () => ({
  useCart: () => ({
    cartItems: mockCartItems,
    cartItemsLength: mockedLength,
    totalCount: mockedCount,
    totalPrice: mockedPrice,
  }),
}));

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('Cart component', () => {
  it('should render Cart component', async () => {
    const { baseElement } = render(<Cart />, { wrapper: MemoryRouter });

    expect(baseElement).toMatchSnapshot();
  });

  it('should render Cart component with 1 count', async () => {
    render(<Cart />, { wrapper: MemoryRouter });

    const result = await screen.findByText('1');
    expect(result).toBeInTheDocument();
  });

  it('should render Cart component with click on icon and get item', async () => {
    render(<Cart />, { wrapper: MemoryRouter });

    fireEvent.click(await screen.findByTestId('cart-icon'));
    expect(await screen.findByText('Ваш заказ')).toBeInTheDocument();
  });

  it("shouldn't render Cart component with non-existent count", async () => {
    render(<Cart />, { wrapper: MemoryRouter });

    const result = screen.queryByText(
      'non-existent description'
    ) as HTMLImageElement;
    expect(result).not.toBeTruthy();
  });
});
