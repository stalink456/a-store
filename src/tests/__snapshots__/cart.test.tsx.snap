// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component should render Cart component 1`] = `
<body>
  <div>
    <div
      class="cart"
    >
      <div
        class="cart__wrapper"
      >
        <button
          class="button__component_j4291 button__ghost_j4291 button__s_j4291 button__component_3xn9f button__ghost_3xn9f cart__icon icon-button__primary_2t17d icon-button__component_2t17d"
          data-testid="cart-icon"
          type="button"
        >
          <span
            class="button__text_j4291 button__stretchText_j4291"
          >
            <span
              class="icon-button__iconWrapper_pctl3 icon-button__s_pctl3"
            >
              <svg
                class="icon-button__icon_pctl3"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M832 312H696v-16c0-101.6-82.4-184-184-184s-184 82.4-184 184v16H192c-17.7 0-32 14.3-32 32v536c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V344c0-17.7-14.3-32-32-32zm-432-16c0-61.9 50.1-112 112-112s112 50.1 112 112v16H400v-16zm392 544H232V384h96v88c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-88h224v88c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-88h96v456z"
                />
              </svg>
            </span>
          </span>
        </button>
        <p
          class="typography__paragraph_c16y5 cart__icon_text typography__primary-medium_c16y5"
        >
          1
        </p>
        <div
          class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as cart__icon_tooltip"
        >
          <div
            class="space__spaceItem_nr3as"
          >
            <p
              class="typography__paragraph_c16y5 cart__icon_tooltip__text typography__primary-large_c16y5 typography__bold_c16y5"
            >
              = 
               â‚½
            </p>
          </div>
        </div>
      </div>
      <div
        class="cart_panel"
      />
    </div>
  </div>
</body>
`;
